src/Symfony/Component/Serializer/Dumper/NormalizerDumper.php

```diff
@@ -0,0 +1,221 @@
+        return <<<EOL
+    public function normalize(\$object, \$format = null, array \$context = array())
+    {
+{$this->generateNormalizeMethodInner($reflectionClass)}
+    }
+EOL;
+    }
+
+    private function generateNormalizeMethodInner(\ReflectionClass $reflectionClass)
+    {
+        $code = <<<EOL
+
+        \$objectHash = spl_object_hash(\$object);
+        if (isset(\$context[ObjectNormalizer::CIRCULAR_REFERENCE_LIMIT][\$objectHash])) {
+            throw new CircularReferenceException('A circular reference has been detected (configured limit: 1).');
+        } else {
+            \$context[ObjectNormalizer::CIRCULAR_REFERENCE_LIMIT][\$objectHash] = 1;
+        }
+
+        \$groups = isset(\$context[ObjectNormalizer::GROUPS]) && is_array(\$context[ObjectNormalizer::GROUPS]) ? \$context[ObjectNormalizer::GROUPS] : null;
+
+        \$output = array();
+EOL;
+
+        $attributesMetadata = $this->classMetadataFactory->getMetadataFor($reflectionClass->name)->getAttributesMetadata();
+        $maxDepthCode = '';
+        foreach ($attributesMetadata as $attributeMetadata) {
+            if (null === $maxDepth = $attributeMetadata->getMaxDepth()) {
+                continue;
+            }
+
+            $key = sprintf(ObjectNormalizer::DEPTH_KEY_PATTERN, $reflectionClass->name, $attributeMetadata->name);
+            $maxDepthCode .= <<<EOL
+
+            if (!isset(\$context['{$key}'])) {
+                \$context['{$key}'] = 1;
+            } else {
+                ++\$context['{$key}'];
+            }
+EOL;
+        }
+
+        if ($maxDepthCode) {
+            $code .= <<<EOL
+
+        if (isset(\$context[ObjectNormalizer::ENABLE_MAX_DEPTH])) {{$maxDepthCode}
+        }
+
+EOL;
+        }
+
+        foreach ($attributesMetadata as $attributeMetadata) {
+            $code .= <<<EOL
+
+        if ((null === \$groups
+EOL;
+
+            if ($attributeMetadata->groups) {
+                $code .= sprintf(" || count(array_intersect(\$groups, array('%s')))", implode("', '", $attributeMetadata->groups));
+            }
+            $code .= ')';
+
+            $code .= " && (!isset(\$context['attributes']) || isset(\$context['attributes']['{$attributeMetadata->name}']) || (is_array(\$context['attributes']) && in_array('{$attributeMetadata->name}', \$context['attributes'], true)))";
+
+            if (null !== $maxDepth = $attributeMetadata->getMaxDepth()) {
+                $key = sprintf(ObjectNormalizer::DEPTH_KEY_PATTERN, $reflectionClass->name, $attributeMetadata->name);
+                $code .= " && (!isset(\$context['{$key}']) || {$maxDepth} >= \$context['{$key}'])";
+            }
+
+            $code .= ') {';
+
+            $value = $this->generateGetAttributeValueExpression($attributeMetadata->name, $reflectionClass);
+            $code .= <<<EOL
+
+            if (is_scalar({$value})) {
+                \$output['{$attributeMetadata->name}'] = {$value};
+            } else {
+                \$subContext = \$context;
```

Comment:
> Should it be used in the below `normalize` call?

Yes indeed. That's strange it isn't covered by the current tests, will have to add one.